### PR TITLE
RHIROS-1122  Create for existing system on Managed Inventory

### PR DESCRIPTION
**Pull request by saltgen**
## Why do we need this change? :thought_balloon:

Scenario:

When ROS is enabled for an existing System on Managed Inventory, the event type is 'updated'.

Changes:
1. Checking for system_metadata along with msg type
2. Assert valid update on inventory consumer with system_metadata

## Documentation update? :memo:

- [ ] Yes
- [x] No

## Security Checklist :lock:

Upon raising this PR please go through [RedHatInsights/secure-coding-checklist](https://github.com/RedHatInsights/secure-coding-checklist)

## :guardsman: Checklist :dart:

- [x] Bugfix
- [ ] New Feature
- [ ] Refactor
- [ ] Unittests Added
- [ ] DRY code
- [ ] Dependency Added
- [ ] DB Migration Added

## Additional :mega:

Tested with local inventory PATCH:
```
inventory-patch-upload:
   curl -X PATCH "[http://localhost:8001/api/inventory/v1/hosts/{1ca44105-16a0-45a7-b33a-d56d781a0c7b}](http://localhost:8001/api/inventory/v1/hosts/%7B1ca44105-16a0-45a7-b33a-d56d781a0c7b%7D)" \
       -H "x-rh-identity: ${b64_identity}" \
       -H "Content-Type: application/json" \
       -H "Accept: application/json" \
       -d '{"display_name": "host1.mydomain.com"}'
```
Update event being handled:

Screenshot from 2023-05-09 18-01-20

### CODE REVIEW

File: ros/processor/inventory_events_consumer.py [hunk 1]
```diff
@@ -119,7 +119,7 @@ def process_system_details(self, msg):
         """ Store new system information (stale, stale_warning timestamp) and return internal DB id"""
         host = msg['host']
         with app.app_context():
-            if msg.get('type') == 'updated':
+            if msg.get("platform_metadata") is None:  # Handles valid update for a ROS system
```
Comment (kgaikwad) [inline ros/processor/inventory_events_consumer.py:122]:
```suggestion
            if msg.get('type') == 'updated' AND msg.get("platform_metadata") is None:  # Handles valid update for a ROS system
```

File: ros/lib/utils.py [hunk 2]
```diff
@@ -268,7 +268,12 @@ def system_allowed_in_ros(msg, reporter):
...
     elif reporter == 'INVENTORY EVENTS':
         cloud_provider = msg['host']['system_profile'].get('cloud_provider')
-        if msg.get('type') == 'updated':
+        # Covers a scenario when ROS is enabled for an existing system on Inventory
+        # i.e. msg type is 'updated' and platform_metadata is available
+        if (
+                msg.get('type') == 'updated'
+                and msg.get('platform_metadata') is None
+        ):
             updated_event = True
```
Comment (kgaikwad) [inline ros/lib/utils.py:277]: For better understanding, could you rename variable `updated_event' to `updated_event_via_api`?

Comment (kgaikwad) [reply ros/lib/utils.py:277]: Btw, I didn't understand why cloud_provider check is ONLY applicable for updated event?

**Review by kgaikwad — CHANGES_REQUESTED** [review]

File: ros/processor/inventory_events_consumer.py (hunk 1 above)
Comment (saltgen) [reply ros/processor/inventory_events_consumer.py:122]: Both conditions are being checked [here](https://github.com/RedHatInsights/ros-backend/pull/324/files#diff-73b796983793ca68d788f97a84b00ef19df5157d5fee7ee854c47ed1133d45f1R274-R275)
I had to add `if msg.get("platform_metadata") is None` again because we are handling create and update in one function

File: ros/lib/utils.py (hunk 2 above)
Comment (saltgen) [reply ros/lib/utils.py]: Ah dang, the cloud_provider check should not be skipped, changes pushed

**Review by kgaikwad — CHANGES_REQUESTED** [review]
@saltgen  - could you please look at this PR that includes nitpicks - https://github.com/saltgen/ros-backend/pull/1/files?
I think, there are multiple unnecessary if..else statements.

**Comment by saltgen** [comment]
> @saltgen - could you please look at this PR that includes nitpicks - https://github.com/saltgen/ros-backend/pull/1/files?
I think, there are multiple unnecessary if..else statements.

@kgaikwad Thanks for the suggestions. I have included all of them here apart from below.

- Redundant parenthesis removed from [here](https://github.com/saltgen/ros-backend/pull/1/files#diff-73b796983793ca68d788f97a84b00ef19df5157d5fee7ee854c47ed1133d45f1R97)

```
>>> x = [1,2,3,4]
j = 5 in x
j = (5 in x)
j
False
j = 5 in x
j
False
```

Restored `and` position [here](https://github.com/saltgen/ros-backend/pull/1/files#diff-73b796983793ca68d788f97a84b00ef19df5157d5fee7ee854c47ed1133d45f1R283)

PEP 8 style guide recommends using a new line after the logical operator i.e. `==`,  `and` to continue the statement onto the next line.

Changes pushed

**Comment by kgaikwad** [comment]
> > @saltgen - could you please look at this PR that includes nitpicks - https://github.com/saltgen/ros-backend/pull/1/files?
> > I think, there are multiple unnecessary if..else statements.
> 
> @kgaikwad Thanks for the suggestions. I have included all of them here apart from below.
> 
>     * Redundant parenthesis removed from [here](https://github.com/saltgen/ros-backend/pull/1/files#diff-73b796983793ca68d788f97a84b00ef19df5157d5fee7ee854c47ed1133d45f1R97)
> 
> 
> ```
> >>> x = [1,2,3,4]
> j = 5 in x
> j = (5 in x)
> j
> False
> j = 5 in x
> j
> False
> ```
> 
>     * Restored `and` position [here](https://github.com/saltgen/ros-backend/pull/1/files#diff-73b796983793ca68d788f97a84b00ef19df5157d5fee7ee854c47ed1133d45f1R283)
> 
> 
> PEP 8 style guide recommends using a new line after the logical operator i.e. `==`, `and` to continue the statement onto the next line.
> 
> Changes pushed

Thanks @saltgen for addressing it! 
Going forward, it will be good to include commit separately and add modifications on top of it (maybe with separate commits) for better understanding.

File: ros/processor/inventory_events_consumer.py (hunk 1 above)
Comment (kgaikwad) [reply ros/processor/inventory_events_consumer.py:122]: I still prefer to add this type check here as well.

Comment (saltgen) [reply ros/processor/inventory_events_consumer.py]: Come to think about it, suggested change is more readable, change pushed

File: ros/processor/inventory_events_consumer.py [hunk 3]
```diff
@@ -119,7 +119,10 @@ def process_system_details(self, msg):
         """ Store new system information (stale, stale_warning timestamp) and return internal DB id"""
         host = msg['host']
         with app.app_context():
-            if msg.get('type') == 'updated':
+            if (
+                    msg.get('type') == 'updated'
+                    and msg.get("platform_metadata") is None  # Handles valid update for a ROS system
+            ):
```
Comment (kgaikwad) [inline ros/processor/inventory_events_consumer.py:125]:
```suggestion
            # 'platform_metadata' field not included when the host is updated via the API.
            if (
                    msg.get('type') == 'updated'
                    and msg.get("platform_metadata") is None
            ):
```

**Review by kgaikwad — APPROVED** [review]
Tests are green! LGTM :+1:

**Comment by saltgen** [comment]
- Tested with all existing archives under `sample_files` dir.
- The conditions seem refined.
- The critical edge case that is targeted in this PR seems covered.

**Note:** 

It would be great if we could get sample msg for an Inventory Update for which we need to create system in ROS, this will be helpful for unit-tests.

**Comment by kgaikwad** [comment]
Thank you @saltgen for verifying change with all sample archives. Merging this change.
